atomic_reactor/cachito_util.py

```diff
@@ -0,0 +1,163 @@
+"""
+Copyright (c) 2019 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+
+from __future__ import absolute_import
+
+from textwrap import dedent
+import json
+import logging
+import time
+
+from atomic_reactor.download import download_url
+from atomic_reactor.util import get_retrying_requests_session
+
+
+logger = logging.getLogger(__name__)
+
+
+class CachitoAPIError(Exception):
+    """Top level exception for errors in interacting with Cachito's API"""
+
+
+class CachitoAPIInvalidRequest(CachitoAPIError):
+    """Invalid request made to Cachito's API"""
+
+
+class CachitoAPIUnsuccessfulRequest(CachitoAPIError):
+    """Cachito's API request not completed successfully"""
+
+
+class CachitpAPIRequestTimeout(CachitoAPIError):
+    """A request to Cachito's API took too long to complete"""
+
+
+class CachitoAPI(object):
+
+    def __init__(self, api_url, insecure=False, token=None, cert=None):
+        self.api_url = api_url
+        self.session = self._make_session(insecure=insecure, cert=cert)
+
+    def _make_session(self, insecure, cert):
+        # method_whitelist=False allows retrying non-idempotent methods like POST
+        session = get_retrying_requests_session(method_whitelist=False)
+        session.verify = not insecure
+        if cert:
+            session.cert = cert
+        return session
+
+    def request_sources(self, repo, ref, flags=None, pkg_managers=None, user=None):
+        """Start a new Cachito request
+
+        :param repo: str, the URL to the SCM repository
+        :param ref: str, the SCM reference to fetch
+        :param pkg_managers: list<str>, list of package managers to be used for resolving
+                             dependencies
+        :param flags: list<str>, list of flag names
+
+        :return: dict, representation of the created Cachito request
+        :raise CachitoAPIInvalidRequest: if Cachito determines the request is invalid
+        """
+        payload = {
+            'repo': repo,
+            'ref': ref,
+            'flags': flags,
+            'pkg_managers': pkg_managers,
+            'user': user,
+        }
+        # Remove None values
+        payload = {k: v for k, v in payload.items() if v}
+
+        url = '{}/api/v1/requests'.format(self.api_url)
+        logger.debug('Making request %s with payload:\n%s', url, json.dumps(payload, indent=4))
+        response = self.session.post(url, json=payload)
+        if response.status_code == 400:
+            raise CachitoAPIInvalidRequest(response.json()['error'])
+        response.raise_for_status()
+        return response.json()
+
+    def wait_for_request(
+            self, request_id, burst_retry=1, burst_length=30, slow_retry=10, timeout=3600):
+        """Wait for a Cachito request to complete
+
+        :param request_id: int, the Cachito request ID
```

Comment:
Would it be nicer if this accepted a Cachito request instead (i.e. the return value from `request_sources`)?

Or we could provide a class which wraps the response and which has a method to get the ID from it?

Or maybe it could accept either?

Same for `download_sources`.

Reply:
Chose to accept either. Code updated.